Clean up exit behavior and error handling

awly opened the pull request:
Users can now choose between os.Exit on errors and handling errors themselves.
Returned error contains usage and an actual error from flag.Parse if that fails.

This is just conceptual approach, perhaps we can utilize FlagSet.Usage better.

Commit:
Return error from ParseStrings for failed flag parsing

ParseStrings should not result in os.Exit for failed flag parsing. Use
flag.ContinueOnError error handling and return error from flagset.Parse
along with usage info if something's wrong.

Also add convenience MustParse function that prints the error and calls
os.Exit(2).

Fixes #3

Commit:
Add tests for invalid input.

Since ParseStrings won't exit on us for invalid flags, we can test some
cases such as mismatched types.

Commit:
Fix double-printing of usage.

Need to set Usage field in FlagSet so that it doesn't print usage on its
own.

cespare commented on flagconf.go:
```diff
@@ -166,6 +186,19 @@ func Parse(path string, config interface{}) error {
+		} else {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		os.Exit(2)
```
Why 2?

awly replied:
Matches behavior of package flag:
http://golang.org/src/flag/flag.go?#L811

cespare replied:
OK thanks

cespare commented on flagconf.go:
```diff
@@ -54,7 +56,25 @@ func ParseStrings(args []string, path string, config interface{}, allowNoConfigF
+		// In case flag parsing fails, return a custom error containing usage
+		// info if user wants to print it.
+		buf := &bytes.Buffer{}
+		flagset.SetOutput(buf)
```
Because this is after `flagset.Parse()`, it won't capture the printing of the error. For instance, this code:

``` go
if err := flagconf.Parse("config.toml", &config); err != nil {
    fmt.Println(err)
}
```

makes this happen:

```
$ ./prog  -asdf
flag provided but not defined: -asdf
flag provided but not defined: -asdf
```

Is that intentional?

cespare commented:
Sorry for the delay on review. This mostly LGTM; just trying to think through the corner cases  (`-h`, `-badflag`).

I gave you commit bit; go ahead and merge when you're happy with this.

awly commented:
No worries, thanks for the review.
Didn't quite test it well, thought that error message was coming from

```
fmt.Fprintln(os.Stderr, ferr.Usage)
```

after type assertion.

Will patch that up in a bit.

awly commented:
@cespare fixed the `-badflag` issue locally.
What do you think is the correct handling of `-h`?
We can add a helper function `IsHelp(error)` so that user doesn't have to do 

```
err.(flagconf.FlagError).Err == flag.ErrHelp
```

or have `FlagError.Error()` append `Usage` to the string in that case.

Commit:
Prevent flagset.ParseStrings from printing error

Set the output for flagset to a bytes.Buffer before calling
flagset.Parse so that it never prints to stderr.

cespare commented:
@alytvynov Another issue is: if you call `MustParse` and then run `prog -h`, it will print just the usage out (but nothing else):

```
$ ./prog -h
  -addr="": listen address (with port)
  -maxprocs=4: maximum OS threads

```

including that extra newline at the end. Instead, I think it should print exactly the same as when you are using `flag.Parse()`.

cespare commented:
@alytvynov `IsHelp` SGTM.

> or have `FlagError.Error()` append `Usage` to the string in that case.

Do you mean prepend?

Commit:
Match help/error output of package flag

Prepend "Usage of progname:\n" to FlagError.Usage and remove trailing
newline. Matches output of flag.Parse on tip exactly.

Commit:
Add IsHelp helper function

Users should use this if they need to detect -h or --help flags, same as
flag.ErrHelp is used.

awly commented:
Ok, i think i got all of them.
Removed trailing newline from `FlagError.Usage` and pre-pended `Usage of <progname>:`.

This matches the output of package flag on tip exactly (r did change some things in flag's output formatting so it may be slightly different from 1.4.2).

IsHelp added too, although i'd expect most users to use MustParse.

PTAL

cespare commented:
Thanks, LGTM. Merged.
